engine/src/test/java/io/zeebe/engine/processor/StreamProcessorHealthTest.java

```diff
@@ -0,0 +1,154 @@
+    streamProcessorRule.writeWorkflowInstanceEventWithSource(ELEMENT_ACTIVATED, 1, firstPosition);
+
+    streamProcessor = getErrorProneStreamProcessor();
+
+    waitUntil(() -> streamProcessor.getHealthStatus() == HealthStatus.HEALTHY);
+    waitUntil(() -> invocation.get() > 1);
+
+    // when
+    streamProcessorRule.getClock().addTime(HEALTH_CHECK_TICK_DURATION.multipliedBy(1));
+    // give some time for scheduled timers to get executed
+    final int retried = invocation.get();
+    waitUntil(() -> retried < invocation.get());
+    streamProcessorRule.getClock().addTime(HEALTH_CHECK_TICK_DURATION.multipliedBy(1));
+
+    // then
+    waitUntil(() -> streamProcessor.getHealthStatus() == HealthStatus.UNHEALTHY);
+  }
+
+  @Test
+  public void shouldMarkUnhealthyWhenInErrorLoop() {
+    // given
+    streamProcessor = getErrorProneStreamProcessor();
+    waitUntil(() -> streamProcessor.getHealthStatus() == HealthStatus.HEALTHY);
+
+    // when
+    streamProcessorRule.writeWorkflowInstanceEvent(ELEMENT_ACTIVATING, 1);
+
+    // then
+    waitUntil(() -> streamProcessor.getHealthStatus() == HealthStatus.UNHEALTHY);
+  }
+
+  @Test
+  public void shouldBecomeHealthyWhenErrorIsResolved() {
+    // given
+    streamProcessor = getErrorProneStreamProcessor();
+    waitUntil(() -> streamProcessor.getHealthStatus() == HealthStatus.HEALTHY);
+    streamProcessorRule.writeWorkflowInstanceEvent(ELEMENT_ACTIVATING, 1);
+    waitUntil(() -> streamProcessor.getHealthStatus() == HealthStatus.UNHEALTHY);
+
+    // when
+    shouldThrowException.set(false);
+
+    // then
+    waitUntil(() -> streamProcessor.getHealthStatus() == HealthStatus.HEALTHY);
+  }
+
+  private StreamProcessor getErrorProneStreamProcessor() {
+    streamProcessor =
+        streamProcessorRule.startTypedStreamProcessor(
+            processingContext -> {
+              mockedZeebeState = spy(processingContext.getZeebeState());
+              processingContext.zeebeState(mockedZeebeState);
+              return processors(mockedZeebeState.getKeyGenerator())
+                  .onEvent(
+                      ValueType.WORKFLOW_INSTANCE,
+                      ELEMENT_ACTIVATING,
+                      new TypedRecordProcessor<>() {
+                        @Override
+                        public void processRecord(
+                            final long position,
+                            final TypedRecord<UnifiedRecordValue> record,
+                            final TypedResponseWriter responseWriter,
+                            final TypedStreamWriter streamWriter,
+                            final Consumer<SideEffectProducer> sideEffect) {
+
+                          invocation.getAndIncrement();
+                          if (shouldThrowException.get()) {
+                            throw new RuntimeException("expected");
+                          }
+                        }
+                      });
+            });
+
+    Mockito.doAnswer(
```

Comment:
This is actually the reason why your test works with the onError loop but this doesn't reflect the real production setting. A this will never fail normally b as I wrote above your logic doesn't detect real loops between write and error or processing and error etc.

Reply:
Yes. I have now added test to cause exception in writeEvent. I couldn't find an easy way to mock failures in updateState though which can also cause an error loop.